### DISCUSSION TITLE
How to enable linting in jupyter notebooks

### DISCUSSION THREAD

**Discussion by f-ritter** [microsoft/vscode-jupyter#10692] [Questions and Answers]
Hello

The readme mentions:
> Editor extensions like [VIM](https://marketplace.visualstudio.com/items?itemName=vscodevim.vim), [bracket colorization](https://marketplace.visualstudio.com/items?itemName=CoenraadS.bracket-pair-colorizer), **linters** and many more are available while editing a cell.

How can we enable linters? The issues don't show up in the problems tab and aren't underlined like in .py files.
It's important to us that we don't have to start this process manually but see issues while coding.

Kind Regards

**Comment by IanMatthewHuff** [comment]
Do you have the "Python" extension also installed? The Jupyter extension doesn't provide the actual linting / formatting support. You need to have the Python extension in addition. If you have that installed you should see linting errors and error squiggles in notebook files.

**Reply by f-ritter** [reply]
Yes, I have Python v2022.8.1, Pylint v2022.2.0 and Jupyter v2022.5.1001601848 installed.
I see the linting errors and squiggles in .py files, but not in .ipynb

**Reply by IanMatthewHuff** [reply]
You have pylint extension installed? Could you look at using the Pylance linter extension instead? I believe that it now ships with Python by default. Linting extensions need special knowledge to provide linting for notebook files and I'm not sure if the Pylint extension has that, but I know that Pylance does. Pylance should be on version 2022.6.30 I believe.

**Reply by f-ritter** [reply]
Had that also installed. Tried to remove pylint to test, but same result.
My coworker has the same issue, btw.

My project settings file looks like this:
```
{
    "python.linting.pylintEnabled": true,
    "python.linting.enabled": true,
    "python.languageServer": "Default"
}
```

**Reply by IanMatthewHuff** [reply]
How about trying the following with the settings. Turn off pylintEnabled. That might be forcing it down the pylint path as opposed to pylance. Then try explicitly setting Python->Language Server to "Pylance". See if that gets it working. Those are the settings that I have.

**Reply by f-ritter** [reply]
Tried that, no success. After that i uninstalled all extensions and reinstalled them again.
I also found these settings which weren't set before:
![image](https://user-images.githubusercontent.com/73236215/177494756-ee607155-38b2-4e65-9f8e-1d321f838711.png)
Still no change. Also tried to remove the project json settings file to see if that might have some negative impact.

My general settings file now looks like this:


```json
{
    "terminal.integrated.defaultProfile.windows": "Command Prompt",
    "git.confirmSync": false,
    "git.autofetch": true,
    "python.linting.pylintEnabled": true,
    "python.linting.flake8Enabled": false,
    "[python]": {
        "editor.defaultFormatter": "ms-python.python"
    },
    "python.languageServer": "Pylance",
    "jupyter.pylanceHandlesNotebooks": true,
    "python.analysis.typeCheckingMode": "strict"
}
```

**Reply by IanMatthewHuff** [reply]
Hmm, so that python.linting.pylintEnabled setting still looks wrong to me. It's true and I'd try that on false to make sure that pylint is not getting started. @debonte Any advice here? I know these settings have been changing around a bit recently.

**Reply by f-ritter** [reply]
Tried it with python.linting.pylintEnabled: false in both settings files also :/

**Reply by debonte** [reply]
@f-ritter, can you provide some sample code along with the errors you are expecting to see (but aren't)?

Pylance is not really a linter. The errors that it reports are more focused on type checking.

**Reply by f-ritter** [reply]
@debonte Oh, okey. Just checked some Pylance scenarios and they're showing. For example:

```python
def example_function(example_text: str):
    print(example_text)

example_function(1)
```
But pylint or flake8 isn't. For example:

```python
a = 1
if a == a:
    a = 2 
```
I should see "Redundant comparison" and "Trailing whitespace" (after a = 2)

**Reply by debonte** [reply]
@karthiknadig, I tried both pylint (pip installed) and the Pylint extension and saw the same behavior as @f-ritter -- we see pylint errors in normal .py files but not in notebooks. Does pylint just not support notebooks at the moment?
